### PR TITLE
Issue 106 fix let underscore

### LINKED ISSUE #106: Replace `let _ =` with either a semicolon if `_` referred to `()` or adding a type signature
see "Returning unit" and "Defensive programming" https://opensource.janestreet.com/standards/ (the latter was clarified at the OCaml meetup to refer to cases where you are applying a function and might accidentally swallow a partially applied function instead of just the value it was supposed to return).

### PR DESCRIPTION
**Pull request by adamhaber**
Removed `let _ =` from most of the files, posted an issue regarding the ones I couldn't. Also - tried to be consistent as possible, but had to deviate here and there. 

Fixes #106

### CODE REVIEW

**Comment by seantalts** [comment]
Oh, would you mind merging master in? And I think I see the issue you were talking about, you posted a comment on #106 right?

**Comment by adamhaber** [comment]
Do you mean merge this to master (on my fork) and then re-submit the PR? And yes, I commented on 106

**Comment by seantalts** [comment]
Assuming you have stan-dev set up as the `upstream` remote and your own fork as `origin`, you can just `git pull upstream master`, fix any conflicts, and then `git push origin issue-106-fix-let-underscore`

**Comment by adamhaber** [comment]
Just to be on the safe side - this is what I did:
```git
git remote add upstream https://github.com/stan-dev/stanc3.git
git pull upstream master
git commit -am "fixed conflict"
git push origin issue-106-fix-let-underscore
```
Tests passed... should by OK now, right?

**Comment by adamhaber** [comment]
Following your reply on the issue, I've added type annotations where replacing with ; didn't work; grep -rn "let _ =" . now returns nothing

File: src/frontend/Semantic_check.ml [hunk 1]
```diff
@@ -1444,7 +1444,7 @@ and semantic_check_fundef_decl ~loc id body =
         if Symbol_table.check_is_unassigned vm id.name then
           error @@ Semantic_error.fn_decl_without_def loc
         else
-          let _ = Symbol_table.set_is_unassigned vm id.name in
+          let () = Symbol_table.set_is_unassigned vm id.name in
```
Comment (seantalts) [inline src/frontend/Semantic_check.ml]: This could also be `else ( Symbol_table.set_is_unassigned vm id.name ; ok () )` but this is fine too

**Pull request merged**